File: raycasting_python/raycaster.py
from pygame.math import Vector2 as vc
from math import cos,sin,tan,atan,sqrt
from settings import *
from stuff import mouse_control

#ray class to cast a single ray

class ray:
    def __init__(self,direction,start_pos) :
        self.pos=start_pos
        self.dir=direction
        
    def cast_ray(self,map,cell_size):
        ray_start=[self.pos.x,self.pos.y]
        ray_step_unit=vc()
        map_cell=[int(ray_start[0]),int(ray_start[1])]
        Ray_length=vc(0,0)
        
        if (self.dir.y<0):
            step_y=-1
            ray_step_unit.y=step_y/self.dir.y
            Ray_length.y=(ray_start[1]-float(map_cell[1]))*ray_step_unit.y
        else :
            step_y=1
            ray_step_unit.y=step_y/self.dir.y
            Ray_length.y=(-ray_start[1]+float(map_cell[1]+1))*ray_step_unit.y
        
        if (self.dir.x<0):
            step_x=-1
            ray_step_unit.x=step_x/self.dir.x
            Ray_length.x=(ray_start[0]-float(map_cell[0]))*ray_step_unit.x
        else :
            step_x=1
            ray_step_unit.x=step_x/self.dir.x
            Ray_length.x=(-ray_start[0]+float(map_cell[0]+1))*ray_step_unit.x
            
        tilefound=False
        distance=0  
        max_distance=10
        while (not tilefound and distance<max_distance):
            if Ray_length.y<Ray_length.x:
                map_cell[1]+=step_y
                distance=Ray_length.y
                Ray_length.y+=ray_step_unit.y
            else:
                map_cell[0]+=step_x
                distance=Ray_length.x
                Ray_length.x+=ray_step_unit.x
            if map[(map_cell[1]//cell_size)][map_cell[0]//cell_size]==1:
                break
        return distance
      
class raycaster():
    def __init__(self,pos,casted_rays):
        self.pos=vc(pos[0]*cell_size,pos[0]*cell_size)
        self.casted_rays=casted_rays
        self.angle=PI
        self.fov=PI/2
        self.step=self.fov/casted_rays
        self.scale=width//casted_rays
        self.sc_distance=half_width/tan(self.fov/2)
        
    def cast_rays(self,map,scale):
        ray_coordonates=[]
        objects=[]
        start_angle=self.angle-self.fov/2
        for rays in range(1,self.casted_rays+1):
            ray_start={'x':self.pos.x/cell_size,'y':self.pos.y/cell_size}
            ray_step_unit=vc()
            map_cell={'x':int(ray_start['x']),'y':int(ray_start['y']) }
            Ray_length=vc(0,0)
            cos_a,sin_a=cos(start_angle),sin(start_angle)
            
            if cos_a<0:
                dx=-1
                ray_step_unit.x=dx/(cos_a)
                Ray_length.x=(ray_start['x']-map_cell['x'])*ray_step_unit.x
            else:
                dx=1
                ray_step_unit.x=dx/(cos_a)
                Ray_length.x=(-ray_start['x']+map_cell['x']+1)*ray_step_unit.x
            if sin_a<0:
                dy=-1
                ray_step_unit.y=dy/(sin_a)
                Ray_length.y=(ray_start['y']-map_cell['y'])*ray_step_unit.y
            else:
                dy=1
                ray_step_unit.y=dy/(sin_a)
                Ray_length.y=(-ray_start['y']+map_cell['y']+1)*ray_step_unit.y 
            while True:
                if Ray_length.x<Ray_length.y:
                    map_cell['x']+=dx
                    dist=Ray_length.x
                    hor=True
                    Ray_length.x+=ray_step_unit.x
                else:
                    map_cell['y']+=dy
                    hor=False
                    dist=Ray_length.y
                    Ray_length.y+=ray_step_unit.y
                if map_cell['x']>=0 and map_cell['x']<len(map) and map_cell['y']>=0 and map_cell['y']<len(map):
                    if map[map_cell['y']][map_cell['x']]!=0:
                        displacement=(ray_start['y']+sin_a*dist)%1 if hor==True else (ray_start['x']+cos_a*dist)%1
                        break
                else:
                    break
            ray_coordonates.append({'x':cos_a*dist*scale,'y':sin_a*dist*scale})
            dist*=cos(start_angle-self.angle)
            projected_height=self.sc_distance/(dist+0.0001)
            objects.append((projected_height,displacement,map[map_cell['y']][map_cell['x']]))
            start_angle=atan(((self.scale*rays-half_width)/self.sc_distance))+self.angle
        
        return objects,ray_coordonates
            
    def update(self,dt,switch,offset,visibility):
        speed=50
        keys=pygame.key.get_pressed()
        cos_a=cos(self.angle)
        sin_a=sin(self.angle)
        vel_v=vc()
        vel_h=vc()
        if keys[pygame.K_z]:
            vel_v.y=sin_a
            vel_v.x=cos_a
        if keys[pygame.K_s]:
            vel_v.y=-sin_a
            vel_v.x=-cos_a
        if keys[pygame.K_d]:
            vel_h.y=cos_a
            vel_h.x=-sin_a
        if keys[pygame.K_q]:
            vel_h.y=-cos_a
            vel_h.x=sin_a
        if keys[pygame.K_UP]:
            offset[0]+=50*dt
        if keys[pygame.K_DOWN]:
            offset[0]-=50*dt
        
        dir=vc((vel_v.x+vel_h.x),(vel_v.y+vel_h.y))
        norm=sqrt((vel_v.x+vel_h.x)*(vel_v.x+vel_h.x)+(vel_v.y+vel_h.y)*(vel_v.y+vel_h.y))
        if norm!=0:
            dir=vc(((vel_v.x+vel_h.x)/norm),((vel_v.y+vel_h.y)/norm))
        self.pos+=dir*dt*speed
        pygame.mouse.set_visible(visibility)
        if switch:
            visibility=False
            mouse_control(self,BORDERS,center,50,0.35,dt,offset)
        if self.angle>=2*PI:
            self.angle=0.01

File: raycasting_c++/src/raycaster.cpp
#include <SFML/Graphics.hpp>
#include <cmath>
#include <iostream>
#include "header_files/raycaster.hpp"

Raycaster::Raycaster(float r, sf::Vector2f position, float fov, int &casted_rays, int width, int height)
    : angle(3.14159f),
      FOV(fov),
      width(width),
      radius(r),
      pos(position)
{
    start_angle = FOV / 2 - (angle);
    scale = int(width / casted_rays);
    shape.setRadius(radius);
    shape.setPosition(pos.x, pos.y);
    shape.setFillColor(sf::Color(255, 0, 0));
    step = (FOV / casted_rays);
}

std::vector<std::vector<float>> Raycaster::update(sf::RenderWindow &screen, double *dt, int cell_size, std::vector<std::vector<int>> &map, int &casted_rays, int &sc_distance)
{
    start_angle = (angle) - (FOV / 2);
    // screen.draw(shape);
    std::vector<std::vector<float>> strips;
    for (int i = 0; i <= casted_rays + 1; i++)
    {
        sf::Vector2f ray_start((pos.x + radius) / cell_size, (pos.y + radius) / cell_size);
        sf::Vector2f unit_step_size, Ray_length;
        int map_cell[] = {int(ray_start.x), int(ray_start.y)};
        bool horizontal, flag = true;
        int dx, dy, index;
        float dist, displacement;
        float cos_a = std::cos(start_angle);
        float sin_a = std::sin(start_angle);
        std::vector<float> data;

        if (cos_a < 0)
        {
            dx = -1;
            unit_step_size.x = dx / cos_a;
            Ray_length.x = (ray_start.x - float(map_cell[0])) * unit_step_size.x;
        }
        else
        {
            dx = 1;
            unit_step_size.x = dx / cos_a;
            Ray_length.x = (float(map_cell[0] + 1) - ray_start.x) * unit_step_size.x;
        }
        if (sin_a < 0)
        {
            dy = -1;
            unit_step_size.y = dy / sin_a;
            Ray_length.y = (ray_start.y - float(map_cell[1])) * unit_step_size.y;
        }
        else
        {
            dy = 1;
            unit_step_size.y = dy / sin_a;
            Ray_length.y = (float(map_cell[1] + 1) - ray_start.y) * unit_step_size.y;
        }
        while (flag)
        {
            if (Ray_length.x < Ray_length.y)
            {
                map_cell[0] += dx;
                dist = Ray_length.x;
                horizontal = true;
                Ray_length.x += unit_step_size.x;
            }
            else
            {
                map_cell[1] += dy;
                horizontal = false;
                dist = Ray_length.y;
                Ray_length.y += unit_step_size.y;
            }
            if (map_cell[0] >= 0 && map_cell[0] < map[0].size() && map_cell[1] >= 0 && map_cell[1] < map.size())
            {
                if (map[map_cell[1]][map_cell[0]] > 0)
                {
                    if (horizontal)
                    {
                        if (cos_a > 0)
                        {
                            displacement = fmod(ray_start.y + sin_a * dist, 1);
                        }
                        else
                        {
                            displacement = 1 - fmod(ray_start.y + sin_a * dist, 1);
                        }
                    }
                    else
                    {
                        if (sin_a > 0)
                        {
                            displacement = fmod(ray_start.x + cos_a * dist, 1);
                        }
                        else
                        {
                            displacement = 1 - fmod(ray_start.x + cos_a * dist, 1);
                        }
                    }
                    index = map[map_cell[1]][map_cell[0]];
                    data.push_back(displacement);
                    data.push_back(float(index));
                    flag = false;
                }
            }
        }
        dist *= std::cos(start_angle - (angle));
        float projected_height = float(sc_distance) / (dist + 0.0001);
        data.push_back(projected_height);
        data.push_back(i);
        start_angle = atan(((scale * i - float(width / 2)) / sc_distance)) + (angle);
        strips.push_back(data);
    }
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Left))
    {
        angle -= 5 * (*dt);
    }
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Right))
    {
        angle += 5 * (*dt);
    }
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Up))
    {
        shape.move(std::cos(angle) * 50 * (*dt), std::sin(angle) * 50 * (*dt));
        pos = shape.getPosition();
    }
    if (sf::Keyboard::isKeyPressed(sf::Keyboard::Down))
    {
        shape.move(-std::cos(angle) * 50 * (*dt), -std::sin(angle) * 50 * (*dt));
        pos = shape.getPosition();
    }
    return strips;
}